src/system.rs

```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+            /// # use uom::si::length::{centimeter, meter};
+            /// # use uom::si::fmt::Arguments;
+            /// # use uom::si::fmt::Indirect;
+            /// # use uom::si::fmt::DisplayStyle::*;
+            /// let length = Length::new::<meter>(1.0);
+            /// let a = Arguments::new(centimeter, Description);
+            /// assert_eq!("100 centimeters", format!("{}", a.with(length)));
+            /// ```
+            /// ## Direct style
+            /// ```
+            /// # use uom::si::Unit;
+            /// # use uom::si::f32::*;
+            /// # use uom::si::length::{centimeter, meter};
+            /// # use uom::si::fmt::Arguments;
+            /// # use uom::si::fmt::Direct;
+            /// # use uom::si::fmt::DisplayStyle::*;
+            /// let length = Length::new::<meter>(1.0);
+            /// let a = Arguments::new(length, centimeter, Description);
+            /// assert_eq!("100 centimeters", format!("{}", a));
+            /// ```
+            #[allow(missing_debug_implementations)] // Prevent accidental direct use.
+            #[derive(Clone, Copy)]
+            pub struct Arguments<T: Unit> {
+                subunit: T,
+                style: DisplayStyle,
+            }
+
+            /// Allows for indirect, delayed formatting of an arbitrary number of quantities
+            /// in the specified style and unit.
+            pub trait Indirect<T>
+            where
+                T: Unit {
+                /// Creates an instance of `Arguments<T>` that can be used to format quantites
+                /// later.
+                ///
+                /// To format a quantity, use the `with` method on `Arguments<T>` to obtain
+                /// a displayable object.
+                fn new(subunit: T, style: DisplayStyle) -> Arguments<T>;
+            }
+
+            /// Allows for direct and immediate formatting of quantities in the specified style
+            /// and unit.
+            pub trait Direct<T>
+            where
+                T: Unit {
+                /// Creates an immediately-displayable object that displays the passed quantity.
+                fn new<D, U, V>(quantity: Quantity<D, U, V>, subunit: T, style: DisplayStyle) -> QuantityArguments<T, D, U, V>
+                where
+                    D: Dimension + ?Sized,
+                    U: Units<V> + ?Sized,
+                    V: Num + Conversion<V>;
+            }
+
+            impl<T: Unit> Direct<T> for Arguments<T>
+            where
+                T: Unit {
+                fn new<D, U, V>(quantity: Quantity<D, U, V>, subunit: T, style: DisplayStyle) -> QuantityArguments<T, D, U, V>
+                where
+                    T: Unit,
+                    D: Dimension + ?Sized,
+                    U: Units<V> + ?Sized,
+                    V: Num + Conversion<V>,
+                    {
+                        let a: Arguments<T> = <Self as Indirect<T>>::new(subunit, style);
```

Comment:
Directly constructor the result? `QuantityArguments { arguments: Arguments { ... }, quantity }`

Reply:
I *believe* the compiler should inline this anyway, and it makes me uneasy to copy-paste code where I could just have a zero-cost call into another entry point. It's unlikely that this struct will need to change, but it's not impossible.